### PR TITLE
[Graph]: Tab Activity Feed Notification Sample code

### PR DESCRIPTION
**Pull request by Trinetra-MSFT**
## Description

This sample demonstrates How to trigger activity feed notification from a Tab application using Microsoft Graph Activity Feed API.
it covers following scenarios
1. S[end notification to user](https://docs.microsoft.com/en-us/graph/api/userteamwork-sendactivitynotification?view=graph-rest-beta&tabs=http)
2. [Send notification to chat](https://docs.microsoft.com/en-us/graph/api/chat-sendactivitynotification?view=graph-rest-beta&tabs=http)
3. [Send notification to team](https://docs.microsoft.com/en-us/graph/api/team-sendactivitynotification?view=graph-rest-beta&tabs=http)

## Testing
1. This sample contains the implementation of various topics as well for all 3 scope
image
image

### CODE REVIEW

File: samples/graph-activity-feed/csharp/ActivityFeedNotification/TabNotificationApp/manifest.json [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        "type": "reservationUpdated",
+        "description": "You have moved up the queue",
+        "templateText": "Here is the {reservationId} with {currentSlot}"
+      }
+    ]
+  },
+  "webApplicationInfo": {
+    "id": "<<Your Microsoft App Id>>",
+    "resource": "<<Your Resource URL>>"
+  }
```
Comment (Wajeed-msft) [inline samples/graph-activity-feed/csharp/ActivityFeedNotification/TabNotificationApp/manifest.json:79]: Check if this is required for current sample.

File: README.md [hunk 2]
```diff
@@ -89,6 +90,7 @@
 |2|Authentication        | Sample illustrating seamless inline authentication for Microsoft Teams apps.                      | | [View][app-auth#ts]
 |3|Complete Samples      | A template for building complex bots (SDK V3) for Microsoft Teams.                                      |[View][app-complete#cs]        |[View][app-complete#ts]
 |4|Meetings Extensibility | Microsoft Teams meeting extensibility sample: token passing |[View][meetings-token-app#cs]     |
+|5|Content Bubble Meetings Extensibility Bot | Microsoft Teams meeting extensibility Content Bubble sample: token passing |[View][apps-in-meeting#cs]     |
```
Comment (Wajeed-msft) [inline README.md:93]: Please remove this. This is not related to Activity feed notification sample.

File: README.md [hunk 3]
```diff
@@ -137,6 +139,8 @@
 
 [meetings-token-app#cs]:samples/meetings-token-app/csharp
 
+[apps-in-meeting#cs]:samples/apps-in-meeting/csharp
```
Comment (Wajeed-msft) [inline README.md:142]: Not related to current sample.

File: samples/graph-activity-feed/csharp/ActivityFeedNotification/Bots/TabActivityFeedBot.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Schema;
+
+namespace Microsoft.BotBuilderSamples.Bots
+{
+    public class TabActivityFeedBot : ActivityHandler
```
Comment (Wajeed-msft) [inline samples/graph-activity-feed/csharp/ActivityFeedNotification/Bots/TabActivityFeedBot.cs:12]: Current sample do not use Bot. Please remove.

File: samples/graph-activity-feed/csharp/ActivityFeedNotification/Controllers/BotController.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Builder.Integration.AspNet.Core;
+
+namespace Microsoft.BotBuilderSamples.Controllers
```
Comment (Wajeed-msft) [inline samples/graph-activity-feed/csharp/ActivityFeedNotification/Controllers/BotController.cs:9]: Not used.

File: samples/graph-activity-feed/csharp/ActivityFeedNotification/Views/Home/Details.cshtml [hunk 6]
```diff
@@ -0,0 +1,30 @@
+﻿
+@model IList<TabActivityFeed.Model.TaskInfo>
```
Comment (Wajeed-msft) [inline samples/graph-activity-feed/csharp/ActivityFeedNotification/Views/Home/Details.cshtml:2]: Check if this is used, if not please delete.

File: samples/graph-activity-feed/csharp/ActivityFeedNotification/Views/Home/GroupChatNotification.cshtml [hunk 7]
```diff
@@ -0,0 +1,231 @@
+﻿
+@model TabActivityFeed.Model.TaskInfo
+@{
+    Layout = "";
+}
+<!DOCTYPE html>
+<html lang="en">
```
Comment (Wajeed-msft) [inline samples/graph-activity-feed/csharp/ActivityFeedNotification/Views/Home/GroupChatNotification.cshtml:7]: This looks like separate cshml page. Please have this in _Layout.cshtml.

File: samples/graph-activity-feed/csharp/ActivityFeedNotification/Views/Home/GroupChatNotification.cshtml [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+    <script src="https://cdn.jsdelivr.net/npm/popper.js@1.16.0/dist/umd/popper.min.js"></script>
+    <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.5.0/js/bootstrap.min.js"></script>
+    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap-select@1.13.14/dist/css/bootstrap-select.min.css">
+    <script src="https://cdn.jsdelivr.net/npm/bootstrap-select@1.13.14/dist/js/bootstrap-select.min.js"></script>
+    <script src="https://statics.teams.microsoft.com/sdk/v1.5.2/js/MicrosoftTeams.min.js" crossorigin="anonymous"></script>
+    <link href="https://cdn.jsdelivr.net/npm/select2@4.1.0-beta.1/dist/css/select2.min.css" rel="stylesheet" />
+    <script src="https://cdn.jsdelivr.net/npm/select2@4.1.0-beta.1/dist/js/select2.min.js"></script>
+    <style>
+        .select2-container {
+            width: inherit !important;
```
Comment (Wajeed-msft) [inline samples/graph-activity-feed/csharp/ActivityFeedNotification/Views/Home/GroupChatNotification.cshtml:24]: Please move this styles to separate file.

File: samples/graph-activity-feed/csharp/ActivityFeedNotification/Views/Home/Index.cshtml [hunk 9]
```diff
@@ -0,0 +1,226 @@
+﻿
+@model TabActivityFeed.Model.TaskInfo
+@{
+    Layout = "";
+}
+<!DOCTYPE html>
+<html lang="en">
+<head>
```
Comment (Wajeed-msft) [inline samples/graph-activity-feed/csharp/ActivityFeedNotification/Views/Home/Index.cshtml:8]: Index.cshtml and GroupChatNotification.cshtml have too much repeat code, please check if it can be combined or moved to partial views

File: samples/graph-activity-feed/csharp/ActivityFeedNotification/appsettings.json [hunk 10]
```diff
@@ -0,0 +1,10 @@
+{
+  "MicrosoftAppId": "",
+  "MicrosoftAppPassword": "",
+  "clientAppId": "",
+  "tenantId": "",
+  "teamId": "",
+  "channelId": "",
+  "UserName": "",
```
Comment (Wajeed-msft) [inline samples/graph-activity-feed/csharp/ActivityFeedNotification/appsettings.json:8]: User name and password should not be in config file. Let's revisit the scenario.

File: samples/graph-activity-feed/csharp/README.md [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+  B) Or from Visual Studio
+
+  - Launch Visual Studio
+  - File -> Open -> Project/Solution
+  - Navigate to `samples/graph-activity-feed/csharp` folder
+  - Select `TabActivityFeed.csproj` file
+  - Press `F5` to run the project
+
+## Testing the bot using Bot Framework Emulator
```
Comment (Wajeed-msft) [inline samples/graph-activity-feed/csharp/README.md:41]: Please correct the readme file. This still talks about Bot testing. Please include details on setting up application permission used in this sample.

**Review by Wajeed-msft — CHANGES_REQUESTED** [review]

File: samples/graph-activity-feed/csharp/ActivityFeedNotification/Views/Home/Details.cshtml (hunk 6 above)
Comment (Trinetra-MSFT) [reply samples/graph-activity-feed/csharp/ActivityFeedNotification/Views/Home/Details.cshtml:2]: Yes this is used as partial view

File: samples/graph-activity-feed/csharp/ActivityFeedNotification/Views/Home/Index.cshtml (hunk 9 above)
Comment (Trinetra-MSFT) [reply samples/graph-activity-feed/csharp/ActivityFeedNotification/Views/Home/Index.cshtml:8]: Both are different one is for static tab (index.cshtml) and other is for chat, I think i can cross verify both and let you know

File: samples/graph-activity-feed/csharp/ActivityFeedNotification/appsettings.json (hunk 10 above)
Comment (Trinetra-MSFT) [reply samples/graph-activity-feed/csharp/ActivityFeedNotification/appsettings.json:8]: Removed from appSetting.json, We have used Username/password auth for Graph API chat message call

File: samples/graph-activity-feed/csharp/README.md [hunk 12]
```diff
@@ -0,0 +1,62 @@
+﻿# Notify Feed App
+
+Bot Framework v4 Activity Feed sample using Tab.
```
Comment (Wajeed-msft) [inline samples/graph-activity-feed/csharp/README.md:3]: This sample does not contain any bot, please correct this line.

File: samples/graph-activity-feed/csharp/README.md [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+  A) From a terminal, navigate to `samples/graph-activity-feed/csharp` folder
+
+  ```bash
+  # run the bot
+  dotnet run
+  ```
+
+  B) Or from Visual Studio
+
+  - Launch Visual Studio
```
Comment (Wajeed-msft) [inline samples/graph-activity-feed/csharp/README.md:35]: This readme file does not talk about setting up Azure AD app which is used. Also there are other parameters in application setting which are not mentioned. 
image

File: samples/graph-activity-feed/csharp/ActivityFeedNotification/Views/Home/Index.cshtml (hunk 9 above)
Comment (Wajeed-msft) [reply samples/graph-activity-feed/csharp/ActivityFeedNotification/Views/Home/Index.cshtml:8]: I still see lot of repeat code. Can we discuss this?

**Review by Wajeed-msft — CHANGES_REQUESTED** [review]

File: samples/graph-activity-feed/csharp/ActivityFeedNotification/Views/Home/Index.cshtml (hunk 9 above)
Comment (Trinetra-MSFT) [reply samples/graph-activity-feed/csharp/ActivityFeedNotification/Views/Home/Index.cshtml:8]: Both view are different and having different methods to be called

File: README.md [hunk 14]
```diff
@@ -22,8 +22,9 @@
 |7|SPFx Tab | Sample app showing Microsoft Teams tabs using SharePoint Framework                                    |   | [View][group-channel-tab#ts#spfx] |
 |8|Tab SSO               | Microsoft Teams sample app for tabs Azure AD SSO                                      | | [View][tab-sso#ts] ,[Teams Toolkit](https://docs.microsoft.com/microsoftteams/platform/toolkit/visual-studio-code-tab-sso)
 |9|Config Tab Authentication               | Microsoft Teams sample app for config tabs Azure AD authentication | [View]()                       | | 
-|10|Deep Link consuming Subentity ID      | Microsoft Teams sample app for demonstrating deeplink from Bot chat to Tab consuming Subentity ID | [View][tab-deeplink#csharp]                       | | [View][tab-deeplink#nodejs]|
-|11|Integrate graph toolkit component in teams tab      | Microsoft Teams tab sample app for demonstrating graph toolkit component |                      | [View][tab-graph-toolkit#js]|
+|10|Tab Activity Feed               | Microsoft Teams sample app for Sending Activity feed notification using Graph API in a Teams Tab. | [View][graph-activity-feed#cs]                       | |
```
Comment (Wajeed-msft) [inline README.md:25]: Could you please move this under graph section?

File: samples/graph-activity-feed/csharp/ActivityFeedNotification/Helpers/ChatMessageHelper.cs [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+        }
+
+        public GraphServiceClient GetGraphClient()
+        {
+            string[] scopes = { "ChannelMessage.Send", "Group.ReadWrite.All", "User.Read" };
+            IPublicClientApplication publicClientApplication = PublicClientApplicationBuilder
+                .Create(_configuration["MicrosoftAppId"])
+                .WithTenantId(_configuration["tenantId"])
+                .Build();
+            UsernamePasswordProvider authenticationProvider = new UsernamePasswordProvider(publicClientApplication, scopes);
```
Comment (Wajeed-msft) [inline samples/graph-activity-feed/csharp/ActivityFeedNotification/Helpers/ChatMessageHelper.cs:202]: Could you please implement authentication instead of using password grant?

File: samples/graph-activity-feed/csharp/ActivityFeedNotification/Helpers/ChatMessageHelper.cs [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+            {
+                Subject = null,
+                Body = new ItemBody
+                {
+                    ContentType = BodyType.Html,
+                    Content = "New Deployment: " + taskInfo.DeployementTitle
+                },
+            };
+            chatMessage.HostedContents = chatMessageHostedContentsCollectionPage;
+            var channelMessage = await graphClientChat.Teams[_configuration["teamId"]].Channels[_configuration["channelId"]].Messages
```
Comment (Wajeed-msft) [inline samples/graph-activity-feed/csharp/ActivityFeedNotification/Helpers/ChatMessageHelper.cs:38]: _configuration["teamId"]] and _configuration["channelId"] are not present in appsettings.json, Is it possible to fetch these dynamically?

File: samples/graph-activity-feed/csharp/ActivityFeedNotification/Views/Shared/TeamNotification.cshtml [hunk 17]
```diff
@@ -0,0 +1,241 @@
+﻿
+@model TabActivityFeed.Model.TaskInfo
+@{
+    Layout = "";
+}
```
Comment (Wajeed-msft) [inline samples/graph-activity-feed/csharp/ActivityFeedNotification/Views/Shared/TeamNotification.cshtml:5]: Why Layout is not used here?

File: samples/graph-activity-feed/csharp/README.md [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+  - File -> Open -> Project/Solution
+  - Navigate to `samples/graph-activity-feed/csharp` folder
+  - Select `TabActivityFeed.csproj` file
+  - Press `F5` to run the project
+
+## Permissions Required to run this sample
+
+  - Following Graph API permissions are required for your Azure App.
+  - `TeamsActivity.Send`, `ChannelMessage.Send`, `ChatMessage.Send`, `Chat.ReadWrite`, `User.Read`,  `TeamsAppInstallation.ReadForUser.All`.
+  - You can navigate to API Permissions section for your App and Choose Microsoft Graph => Application/Delegated permissions.
```
Comment (Wajeed-msft) [inline samples/graph-activity-feed/csharp/README.md:45]: Do we need application permissions / delegated ones?

**Review by Wajeed-msft — CHANGES_REQUESTED** [review]

**Review by Wajeed-msft — APPROVED** [review]

**Review by asnow003 — APPROVED** [review]
